presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragment.java

```diff
@@ -51,6 +52,7 @@
     private final List<RemoteSourceNode> remoteSourceNodes;
     private final PartitioningScheme partitioningScheme;
     private final StageExecutionStrategy stageExecutionStrategy;
+    private final StatsAndCosts statsAndCosts;
```

Comment:
Why is this segmented per fragment? The costs are per node, so they should be associated with the overall query plan.

Reply:
We had to attach the stats to the `Fragment`, so it is easier to pass it to the `QueryMonitor` and `PlanPrinter`. 

Otherwise you need to toss stats around across many classes. `SqlQueryExecution#analyzeQuery` -> `PlanRoot` -> `SqlQueryExecution#planDistribution` -> `SqlQueryScheduler` -> `StageInfo` -> `QueryInfo` -> `QueryMonitor` 

It can be simplified somehow, by taking the stats in the `SqlQueryExecution#buildQueryInfo` from the `SqlQueryExecution#queryPlan`, and storing the stats directly into the `QueryInfo`. But than the stats object has to be tossed around in the `PlanPrinter`.

Reply:
they are associated with the overall plan, but once you have a list of fragments, you need to still be able to get the stats and costs